api/observability/v1/clusterlogforwarder_types.go

```diff
@@ -0,0 +1,190 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package v1
+
+import (
+	"github.com/openshift/cluster-logging-operator/internal/status"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ClusterLogForwarderSpec defines the desired state of ClusterLogForwarder
+type ClusterLogForwarderSpec struct {
+	// Inputs are named filters for log messages to be forwarded.
+	//
+	// There are three built-in inputs named `application`, `infrastructure` and
+	// `audit`. You don't need to define inputs here if those are sufficient for
+	// your needs. See `inputRefs` for more.
+	//
+	// +optional
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Forwarder Inputs",xDescriptors={"urn:alm:descriptor:com.tectonic.ui:forwarderInputs"}
+	Inputs []InputSpec `json:"inputs,omitempty"`
+
+	// Outputs are named destinations for log messages.
+	//
+	// There is a built-in output named `default` which forwards to the default
+	// openshift log store. You can define outputs to forward to other stores or
+	// log processors, inside or outside the cluster.
+	//
+	// +optional
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Forwarder Outputs",xDescriptors={"urn:alm:descriptor:com.tectonic.ui:forwarderOutputs"}
+	Outputs []OutputSpec `json:"outputs,omitempty"`
+
+	// Filters are applied to log records passing through a pipeline.
+	// There are different types of filter that can select and modify log records in different ways.
+	// See [FilterTypeSpec] for a list of filter types.
+	// +optional
+	Filters []FilterSpec `json:"filters,omitempty"`
+
+	// Pipelines forward the messages selected by a set of inputs to a set of outputs.
+	//
+	// +required
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Forwarder Pipelines",xDescriptors={"urn:alm:descriptor:com.tectonic.ui:forwarderPipelines"}
+	Pipelines []PipelineSpec `json:"pipelines,omitempty"`
```

Comment:
The `xDescriptors` seem not to be correct here. AFAIU the only allowed descriptors are given here: https://github.com/openshift/console/blob/master/frontend/packages/operator-lifecycle-manager/src/components/descriptors/reference/reference.md

Reply:
Also all four `Filters, `Inputs`, `Outputs` and `Pipelines` should be annotated with the following markers to make them compatible with server-side-apply:
- listType: `//+listType:=map`
- listMapKey:  `//+listMapKey:=name"

This allows the APIserver to enable ownership of each list item in inputs/outputs/pipelines independently because each list iteam can be uniquely identified by the field `name`. This enables granular use as proposed in the [LoggingStack proposal](https://github.com/openshift/enhancements/pull/1566/files#diff-103a5e73dc036b3386cc6a886af843c7833bf90ebbbe6583bd9dc62172a7326dR247-R258)